File: legacy-python/engine.py
"""Offline quiz rules; vocabulary and synonym groups preserved from v2."""
from dataclasses import dataclass
from pathlib import Path
import random
import time

POS={'noun':'名詞','verb':'動詞','adjective':'形容詞','adverb':'副詞'}
LEVELS={'basic':'基礎編','advanced':'応用編','upper1':'上級Ⅰ','upper2':'上級Ⅱ'}
MODES={'random':'ランダム','weak':'苦手優先','new':'未学習のみ','due':'復習日が来た単語'}

@dataclass(frozen=True)
class Word:
    english:str
    japanese:str
    level:str
    pos:str
    @property
    def key(self):
        return f'{self.english}:{self.pos}'

def load_words():
    root=Path(__file__).parent
    words=[]
    for name in ('vocabulary.txt','vocabulary_upper.txt'):
        level=pos=None
        for line in (root/name).read_text(encoding='utf-8').splitlines():
            line=line.strip()
            if not line or line.startswith('#'):
                continue
            if line.startswith('['):
                level,pos=line.strip('[]').split(',')
                if level not in LEVELS or pos not in POS:
                    raise ValueError('Invalid vocabulary section')
            else:
                english,japanese=line.split('=',1)
                if not english or not japanese or level is None:
                    raise ValueError('Invalid vocabulary row')
                words.append(Word(english,japanese,level,pos))
    if len({w.english for w in words})!=2000:
        raise ValueError('Expected 2000 unique words')
    for key in LEVELS:
        if sum(w.level==key for w in words)!=500:
            raise ValueError('Expected 500 words per course')
    return words

class QuestionFactory:
    def __init__(self,words,rng=None):
        self.words=words
        self.rng=rng or random.Random()
        self.previous={}
        self.levels=set(LEVELS)
        self.family_ids={}
        for i,line in enumerate(Path(__file__).with_name('synonyms.txt').read_text(encoding='utf-8').splitlines()):
            for english in line.split():
                self.family_ids.setdefault(english,set()).add(i)
    def compatible(self,a,b):
        return (a.pos==b.pos and a.english!=b.english
            and a.japanese not in b.japanese and b.japanese not in a.japanese
            and not self.family_ids.get(a.english,set()).intersection(self.family_ids.get(b.english,set())))
    def options(self,target):
        pool=[w for w in self.words if w.level in self.levels and self.compatible(target,w)]
        same=[w for w in pool if w.level==target.level]
        other=[w for w in pool if w.level!=target.level]
        for attempt in range(100):
            self.rng.shuffle(same)
            self.rng.shuffle(other)
            choices=[target]
            for candidate in same+other:
                if all(self.compatible(candidate,w) for w in choices):
                    choices.append(candidate)
                if len(choices)==6:
                    break
            signature=frozenset(w.key for w in choices)
            if len(choices)==6 and signature!=self.previous.get(target.key):
                self.previous[target.key]=signature
                self.rng.shuffle(choices)
                return choices
        raise ValueError('Cannot create six choices for '+target.english)

@dataclass(frozen=True)
class Answer:
    word:Word
    selected:Word|None
    elapsed:float
    @property
    def correct(self):
        return self.word==self.selected

class Session:
    def __init__(self,words,factory,seconds,clock=time.monotonic):
        if not words or not 3<=seconds<=120:
            raise ValueError('Invalid session settings')
        self.words=list(words)
        self.factory=factory
        self.seconds=seconds
        self.clock=clock
        self.index=-1
        self.answers=[]
        self.answered=True
        self.finished=False
        self.options=[]
        self.deadline=0
    def next(self):
        if not self.answered or self.finished:
            return False
        if self.index+1==len(self.words):
            self.finished=True
            return False
        self.index+=1
        self.word=self.words[self.index]
        self.options=self.factory.options(self.word)
        self.started=self.clock()
        self.deadline=self.started+self.seconds
        self.answered=False
        return True
    @property
    def remaining(self):
        return max(0,self.deadline-self.clock())
    def answer(self,index=None):
        if self.answered or self.finished or self.index<0:
            return None
        if index is not None and (type(index) is not int or not 0<=index<6):
            raise ValueError('Invalid choice')
        elapsed=max(0,self.clock()-self.started)
        selected=None if index is None or elapsed>=self.seconds else self.options[index]
        result=Answer(self.word,selected,min(self.seconds,elapsed))
        self.answers.append(result)
        self.answered=True
        return result
    @property
    def mistakes(self):
        return [a.word for a in self.answers if not a.correct]

File: src/engine.ts
import {
  type Word,
  type Level,
  type Mode,
  type Snapshot,
  type Progress,
  type Attempt,
} from './model.js';
export type Random = () => number;
export function shuffle<T>(
  items: readonly T[],
  rng: Random = Math.random,
): T[] {
  const a = [...items];
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}
export function eligible(
  words: Word[],
  levels: Level[],
  mode: Mode,
  p: Snapshot,
  now = Date.now(),
): Word[] {
  return words.filter(
    (w) =>
      levels.includes(w.level) &&
      (mode === 'new'
        ? !p[w.key]
        : mode === 'due'
          ? !!p[w.key] && p[w.key].dueAt <= now
          : true),
  );
}
export function priority(w: Word, p: Snapshot): number[] {
  const v = p[w.key];
  return v && v.seen > v.correct && v.streak < 3
    ? [0, v.streak, v.correct / v.seen]
    : [v ? 2 : 1, 0, 0];
}
export function selectWords(
  pool: Word[],
  count: number,
  mode: Mode,
  p: Snapshot,
  rng: Random = Math.random,
): Word[] {
  if (!Number.isInteger(count) || count < 1 || count > pool.length)
    throw new Error(`問題数は1〜${pool.length}問にしてください。`);
  let a = shuffle(pool, rng);
  if (mode === 'weak') {
    a.sort((x, y) => {
      const u = priority(x, p),
        v = priority(y, p);
      return u[0] - v[0] || u[1] - v[1] || u[2] - v[2];
    });
    a = shuffle(a.slice(0, count), rng);
  }
  return a.slice(0, count);
}
export function goalWords(
  pool: Word[],
  count: number,
  today: Set<string>,
  rng: Random = Math.random,
): Word[] {
  if (!Number.isInteger(count) || count < 1 || count > pool.length)
    throw new Error(
      `コースは${count}問、選択範囲は${pool.length}語です。範囲を増やすか目標を下げてください。`,
    );
  const fresh = shuffle(
    pool.filter((w) => !today.has(w.key)),
    rng,
  ).slice(0, count);
  return shuffle(
    [
      ...fresh,
      ...shuffle(
        pool.filter((w) => today.has(w.key)),
        rng,
      ).slice(0, count - fresh.length),
    ],
    rng,
  );
}
export class QuestionFactory {
  previous = new Map<string, string>();
  families = new Map<string, Set<number>>();
  constructor(
    public words: Word[],
    families: string[][],
    public rng: Random = Math.random,
  ) {
    families.forEach((group, i) =>
      group.forEach((word) => {
        if (!this.families.has(word)) this.families.set(word, new Set());
        this.families.get(word)!.add(i);
      }),
    );
  }
  compatible(a: Word, b: Word): boolean {
    return (
      a.pos === b.pos &&
      a.english !== b.english &&
      !a.japanese.includes(b.japanese) &&
      !b.japanese.includes(a.japanese) &&
      ![...(this.families.get(a.english) || [])].some((id) =>
        this.families.get(b.english)?.has(id),
      )
    );
  }
  options(target: Word, levels: Level[]): Word[] {
    const pool = this.words.filter(
      (w) => levels.includes(w.level) && this.compatible(target, w),
    );
    const same = pool.filter((w) => w.level === target.level),
      other = pool.filter((w) => w.level !== target.level);
    for (let attempt = 0; attempt < 100; attempt++) {
      const choices = [target];
      for (const candidate of [
        ...shuffle(same, this.rng),
        ...shuffle(other, this.rng),
      ]) {
        if (choices.every((w) => this.compatible(candidate, w)))
          choices.push(candidate);
        if (choices.length === 6) break;
      }
      const signature = choices
        .map((w) => w.key)
        .sort()
        .join('|');
      if (choices.length === 6 && signature !== this.previous.get(target.key)) {
        this.previous.set(target.key, signature);
        return shuffle(choices, this.rng);
      }
    }
    throw new Error(`${target.english}の選択肢を作成できませんでした。`);
  }
}
export interface Answer {
  word: Word;
  selected: Word | null;
  correct: boolean;
  elapsed: number;
}
export class Quiz {
  index = -1;
  answers: Answer[] = [];
  answered = true;
  finished = false;
  options: Word[] = [];
  started = 0;
  deadline = 0;
  constructor(
    public words: Word[],
    public factory: QuestionFactory,
    public seconds: number,
    public levels: Level[],
    public clock = () => Date.now() / 1000,
  ) {
    if (!words.length || seconds < 3 || seconds > 120)
      throw new Error('制限時間は3〜120秒です。');
  }
  get word() {
    return this.words[this.index];
  }
  get remaining() {
    return Math.max(0, this.deadline - this.clock());
  }
  get mistakes() {
    return this.answers.filter((a) => !a.correct).map((a) => a.word);
  }
  next() {
    if (!this.answered || this.finished) return false;
    if (this.index + 1 === this.words.length) {
      this.finished = true;
      return false;
    }
    this.index++;
    this.options = this.factory.options(this.word, this.levels);
    this.started = this.clock();
    this.deadline = this.started + this.seconds;
    this.answered = false;
    return true;
  }
  answer(index: number | null): Answer | null {
    if (this.answered || this.finished || this.index < 0) return null;
    if (index !== null && (!Number.isInteger(index) || index < 0 || index > 5))
      throw new Error('選択肢が不正です。');
    const elapsed = Math.max(0, this.clock() - this.started);
    const selected =
      index === null || elapsed >= this.seconds ? null : this.options[index];
    const result = {
      word: this.word,
      selected,
      correct: selected?.key === this.word.key,
      elapsed: Math.min(this.seconds, elapsed),
    };
    this.answers.push(result);
    this.answered = true;
    return result;
  }
}
export function advance(
  old: Progress | undefined,
  word: string,
  correct: boolean,
  at: number,
): Progress {
  const streak = correct ? (old?.streak || 0) + 1 : 0;
  return {
    word,
    seen: (old?.seen || 0) + 1,
    correct: (old?.correct || 0) + Number(correct),
    streak,
    lastAt: at,
    dueAt:
      at +
      (correct
        ? [1, 3, 7, 14, 30, 60][Math.min(streak - 1, 5)] * 86400000
        : 600000),
  };
}
export function snapshot(attempts: Attempt[]): Snapshot {
  const p: Snapshot = Object.create(null);
  for (const a of attempts)
    p[a.word] = advance(p[a.word], a.word, a.correct, a.at);
  return p;
}
// Python round(): ties to even, including e.g. 1/8 = 12.5% -> 12%.
export function roundEven(value: number) {
  const n = Math.floor(value),
    f = value - n;
  return f === 0.5 ? n + (n % 2) : Math.round(value);
}
export function accuracy(correct: number, total: number) {
  return total ? `${roundEven((correct / total) * 100)}%` : '—';
}
export function dayKey(date = new Date()): string {
  return `${date.getFullYear().toString().padStart(4, '0')}-${String(date.getMonth() + 1).padStart(2, '0')}-${String(date.getDate()).padStart(2, '0')}`;
}
export function daySummary(attempts: Attempt[], day = dayKey()) {
  const rows = attempts.filter((a) => dayKey(new Date(a.at)) === day);
  const keys = new Set(rows.map((a) => a.word));
  return {
    words: keys.size,
    answers: rows.length,
    correct: rows.filter((a) => a.correct).length,
    keys,
  };
}
export function monthSummary(attempts: Attempt[], year: number, month: number) {
  const prefix = `${year.toString().padStart(4, '0')}-${String(month + 1).padStart(2, '0')}-`;
  const rows = attempts.filter((a) =>
    dayKey(new Date(a.at)).startsWith(prefix),
  );
  const days: Record<string, ReturnType<typeof daySummary>> = Object.create(
    null,
  );
  const groups = new Map<string, Attempt[]>();
  for (const a of rows) {
    const k = dayKey(new Date(a.at));
    if (!groups.has(k)) groups.set(k, []);
    groups.get(k)!.push(a);
  }
  for (const [key, list] of groups) days[key] = daySummary(list, key);
  return {
    days,
    total: Object.values(days).reduce((s, d) => s + d.words, 0),
    unique: new Set(rows.map((a) => a.word)).size,
    answers: rows.length,
    activeDays: groups.size,
  };
}
export function streak(attempts: Attempt[], today = new Date()) {
  const active = new Set(attempts.map((a) => dayKey(new Date(a.at))));
  const day = new Date(
    today.getFullYear(),
    today.getMonth(),
    today.getDate(),
    12,
  );
  if (!active.has(dayKey(day))) day.setDate(day.getDate() - 1);
  let n = 0;
  while (active.has(dayKey(day))) {
    n++;
    day.setDate(day.getDate() - 1);
  }
  return n;
}
